include/boost/math/fft/abstract_ring.hpp

```diff
@@ -50,6 +51,27 @@
   */
 
   namespace boost { namespace math {  namespace fft {  namespace detail {
+  
+    #if __cplusplus < 201700L
+    template<typename ...> using void_t = void;
+    #else
+    using std::void_t;
+    #endif
+    
+
+    template<class T, typename = void_t<> >
+    struct is_complex : std::false_type
+    {};
+
+    template<class T >
+    struct is_complex<T, 
+        void_t<
+            typename T::value_type,
+            decltype(sin(std::declval<typename T::value_type>())),
+            decltype(cos(std::declval<typename T::value_type>()))
+         > > : std::true_type
```

Comment:
It's a very nice template. However I am a little bit concerned with it's name. Maybe better to be more explicit that it checks for presence of `sin(…)` and `cos(…)`. Like maybe `has_sin_cos<…>` ?

Following this, I think we should to go through Boost libs and add tests for stuff which happens to have `sin(…)` and `cos(…)`, [this one](https://www.boost.org/doc/libs/1_76_0/libs/math/doc/html/math_toolkit/trans.html) is the first contender.

It's also possible to calculate the `sin(…)` and `cos(…)` from a matrix.

Reply:
So far we have two kinds of template algorithms: 
1. those that assume the type is complex so that we can use sin/cos to compute the roots of unity (so that we don't have to ask for the user for the roots and we don't have to use power functions to derive smaller roots from larger ones);
2. and those who would work even in a larger domain (this domain include complex as well) of algebraic ring types.

The intention of this template is to tell if the argument represents a complex number, so that we can choose one DFT implementation or the other. Checking for sin/cos is a first approach to deduce the answer for unknown types. We could change the logic if needed to.

Reply:
OK, I see. Then maybe simpler it would be to use the native `boost::is_complex` from `#include <boost/type_traits/is_complex.hpp>` ?

Reply:
That would be great! But the documentation says
*If T is a complex number type then true (of type std::complex<U> for some type U), otherwise false. *

Reply:
Sure, we could alias and specialize the template `boost::is_complex` inside the `boost::math::fft` namespace to evaluate to `true` all `boost::multiprecision::cpp_complex<...>` types. But then I guess `boost::math` would be dependent on `boost::type_traits` and `boost::multiprecision`.

Reply:
> That means (I've just tested) `boost::is_complex<boost::multiprecision::cpp_complex_quad>::value` is false.

@ckormanyos , this seems like a bug to me, what do you think?

Reply:
As for the **Type Traits** library, that's not a bug
https://github.com/boostorg/type_traits/blob/develop/include/boost/type_traits/is_complex.hpp
The value is true if the type is `std::complex<...>` otherwise false. Hence the template tells if the type is "complex", where "complex" means `std::complex`.

**Multiprecision** has its own definition of the `is_complex` template
https://github.com/boostorg/multiprecision/blob/develop/include/boost/multiprecision/traits/is_complex.hpp 
But it does the same thing.

Reply:
Thanks for checking this. We will have time figure this out. Maybe for example check if the type uses `complex_adaptor<…>`. But it's not urgent. What matters is that now I understand your intentions in this part of code :)

Reply:
In `complex_adaptor`, we use things like

enable or test a Boolean value from:
```
boost::multiprecision::detail::is_complex<Result>::value
```

I need to check again if this works for multiprecision types.

But here is a rather restricting constraint. Very much work was done to make Boost.Math standalone, requiring no other Boost library dependencies other than itself. It might be incongruent with the evolution of Boost.Math to have FFT depend on Boost.Multiprecision.

Reply:
> .. to have FFT depend on Boost.Multiprecision.

I definitely agree with that. We will need to do more SFINAE tricks to make sure that a simple line that calls FFT on a simple double type does not pull extra dependencies with it.